lib/ansible/modules/windows/win_domain_group_membership.ps1

```diff
@@ -0,0 +1,155 @@
+}
+
+$result = @{
+    changed = $false
+    name = $name
+}
+if ($state -in @("present", "pure")) {
+    $result.added = @()
+}
+if ($state -in @("absent", "pure")) {
+    $result.removed = @()
+}
+if ($diff_mode) {
+    $result.diff = @{}
+}
+
+try {
+    Get-ADGroup -Identity $name -Properties * @extra_args
+}
+catch {
+    Fail-Json -obj $result "Could not find domain group $name"
+}
+
+$current_members = Get-AdGroupMember -Identity $name @extra_args
+$pure_members = @()
+
+foreach ($member in $members) {
+    try {
+        $group_member = Get-ADUser -Identity $member @extra_args
+    }
+    catch {
+        $group_member = $null
+    }
+    if (!$group_member) {
+       try {
+            $group_member = Get-ADGroup -Identity $member @extra_args
+        }
+        catch {
+            $group_member = $null
+        }
+    }
+    if (!$group_member) {
+        Fail-Json -obj $result "Could not find domain user or group $member"
+    }
+
+    if ($state -eq "pure") {
+        $pure_members += $group_member
+    }
+
+    $user_in_group = $false
+    foreach ($current_member in $current_members) {
+        if ($current_member.sid -eq $group_member.sid) {
+            $user_in_group = $true
+            break
+        }
+    }
+
+    try {
+        if ($state -in @("present", "pure") -and !$user_in_group) {
+            Add-ADGroupMember -Identity $name -Members $group_member -WhatIf:$check_mode @extra_args
+            $result.added += $group_member.SamAccountName
+            $result.changed = $true
+        } elseif ($state -eq "absent" -and $user_in_group) {
+            Remove-ADGroupMember -Identity $name -Members $group_member -WhatIf:$check_mode @extra_args
+            $result.removed += $group_member.SamAccountName
+            $result.changed = $true
+        }
+    } catch {
+        Fail-Json -obj $result -message $_.Exception.Message
+    }
+}
+
+if ($state -eq "pure") {
+    # Perform removals for existing group members not defined in $members
+    $current_members = Get-AdGroupMember -Identity $name @extra_args
+
+    foreach ($current_member in $current_members) {
+        $user_to_remove = $true
+        foreach ($pure_member in $pure_members) {
+            if ($pure_member.sid -eq $current_member.sid) {
+                $user_to_remove = $false
+                break
+            }
+        }
+
+        try {
+            if ($user_to_remove) {
+                Remove-ADGroupMember -Identity $name -Members $group_member -WhatIf:$check_mode @extra_args
+                $result.removed += $current_member.SamAccountName
+                $result.changed = $true
+            }
+        } catch {
+            Fail-Json -obj $result -message $_.Exception.Message
+        }
+    }
+}
+
+$final_members = Get-AdGroupMember -Identity $name @extra_args
+
+if ($final_members) {
+    $result.members = [Array]$final_members.name
+} else {
+    $result.members = @()
+}
+
+if ($diff_mode -and $result.changed) {
```

Comment:
Using the `prepared` part of a diff should be avoided. If you use the before/after keys under diff Ansible will generate a nicer diff output with colour, unlike prepared.

Because of this I would do something like

```
$result.diff.before = @(list of members originally in the group)
$result.diff.after = @(list of member now in the group after module execution)
```

The controller side code will handle the + and - elements for you based on the list diffs.

Reply:
A List did not work for me. It complained about not finding split on the list. But changed from `prepared ` to `before`/`after` for the diff.

Reply:
Hmm ok, I would need to test out that, I've only really done a dict that contained a list with the values and not a list directly. Thanks for letting me know.